File: backend/src/services/chat.service.ts
import mongoose from 'mongoose';
import { Chat, IChat, IChatMessage } from '../models/Chat';
import { Order } from '../models/Order';
import { Customer } from '../models/Customer';
import { Restaurant } from '../models/Restaurant';
import { Driver } from '../models/Driver';
import { NotFoundError, BadRequestError, ForbiddenError } from '../utils/errors';
import { emitToUser, emitToRestaurant, emitToDriver, emitToOrder } from '../config/socket';

class ChatService {
  /**
   * Get or create a chat for an order
   */
  async getOrCreateChat(
    orderId: string,
    chatType: 'customer_restaurant' | 'customer_driver' | 'restaurant_driver',
    userId: string,
    userRole: string
  ): Promise<IChat> {
    const order = await Order.findById(orderId);
    if (!order) {
      throw new NotFoundError('الطلب غير موجود');
    }

    // Verify user is a participant
    const isParticipant = await this.verifyParticipant(order, userId, userRole, chatType);
    if (!isParticipant) {
      throw new ForbiddenError('غير مصرح لك بالوصول لهذه المحادثة');
    }

    // Get participant IDs
    const participants = {
      customerId: order.customerId,
      restaurantId: undefined as mongoose.Types.ObjectId | undefined,
      driverId: undefined as mongoose.Types.ObjectId | undefined,
    };

    if (chatType === 'customer_restaurant' || chatType === 'restaurant_driver') {
      const restaurant = await Restaurant.findById(order.restaurantId);
      if (restaurant) {
        participants.restaurantId = restaurant._id as mongoose.Types.ObjectId;
      }
    }

    if (chatType === 'customer_driver' || chatType === 'restaurant_driver') {
      if (order.driverId) {
        const driver = await Driver.findOne({ userId: order.driverId });
        if (driver) {
          participants.driverId = driver._id as mongoose.Types.ObjectId;
        }
      }
    }

    // Find or create the chat
    let chat = await Chat.findOne({ orderId, chatType });

    if (!chat) {
      chat = await Chat.create({
        orderId,
        chatType,
        participants,
        messages: [],
        isActive: true,
      });
    }

    return chat;
  }

  /**
   * Send a message in a chat
   */
  async sendMessage(
    chatId: string,
    userId: string,
    userRole: string,
    content: string,
    type: 'text' | 'image' | 'location' = 'text',
    extras?: { imageUrl?: string; location?: { lat: number; lng: number } }
  ): Promise<IChatMessage> {
    const chat = await Chat.findById(chatId);
    if (!chat) {
      throw new NotFoundError('المحادثة غير موجودة');
    }

    if (!chat.isActive) {
      throw new BadRequestError('المحادثة مغلقة');
    }

    // Determine sender role for chat
    let senderRole: 'customer' | 'restaurant' | 'driver';
    if (userRole === 'customer') {
      senderRole = 'customer';
    } else if (userRole === 'restaurant') {
      senderRole = 'restaurant';
    } else if (userRole === 'driver' || userRole === 'delivery') {
      senderRole = 'driver';
    } else {
      throw new ForbiddenError('غير مصرح لك بإرسال رسائل');
    }

    // Add message
    const message: IChatMessage = {
      sender: new mongoose.Types.ObjectId(userId),
      senderRole,
      content,
      type,
      imageUrl: extras?.imageUrl,
      location: extras?.location,
      isRead: false,
      createdAt: new Date(),
    };

    chat.messages.push(message);
    chat.lastMessage = type === 'text' ? content : type === 'image' ? '📷 صورة' : '📍 موقع';
    chat.lastMessageAt = new Date();
    chat.lastMessageBy = new mongoose.Types.ObjectId(userId);

    await chat.save();

    const newMessage = chat.messages[chat.messages.length - 1];

    // Emit socket event for real-time updates
    const messageData = {
      chatId: chat._id,
      orderId: chat.orderId,
      message: newMessage,
    };

    // Notify participants
    if (chat.participants.customerId && senderRole !== 'customer') {
      const customer = await Customer.findById(chat.participants.customerId);
      if (customer) {
        emitToUser(customer.userId.toString(), 'chat:message', messageData);
      }
    }

    if (chat.participants.restaurantId && senderRole !== 'restaurant') {
      emitToRestaurant(chat.participants.restaurantId.toString(), 'chat:message', messageData);
    }

    if (chat.participants.driverId && senderRole !== 'driver') {
      const driver = await Driver.findById(chat.participants.driverId);
      if (driver) {
        emitToDriver(driver.userId.toString(), 'chat:message', messageData);
      }
    }

    // Also emit to order room
    emitToOrder(chat.orderId.toString(), 'chat:message', messageData);

    return newMessage;
  }

  /**
   * Get chat messages with pagination
   */
  async getMessages(
    chatId: string,
    userId: string,
    userRole: string,
    page: number = 1,
    limit: number = 50
  ): Promise<{
    messages: IChatMessage[];
    total: number;
    page: number;
    pages: number;
  }> {
    const chat = await Chat.findById(chatId);
    if (!chat) {
      throw new NotFoundError('المحادثة غير موجودة');
    }

    // Mark messages as read
    await chat.markAsRead(new mongoose.Types.ObjectId(userId));

    // Get paginated messages (most recent first)
    const total = chat.messages.length;
    const pages = Math.ceil(total / limit);
    const startIndex = Math.max(0, total - page * limit);
    const endIndex = total - (page - 1) * limit;

    const messages = chat.messages
      .slice(startIndex, endIndex)
      .reverse(); // Return in chronological order

    return {
      messages,
      total,
      page,
      pages,
    };
  }

  /**
   * Mark chat messages as read
   */
  async markAsRead(chatId: string, userId: string): Promise<boolean> {
    const chat = await Chat.findById(chatId);
    if (!chat) {
      throw new NotFoundError('المحادثة غير موجودة');
    }

    const updated = await chat.markAsRead(new mongoose.Types.ObjectId(userId));

    if (updated) {
      // Emit read receipt
      emitToOrder(chat.orderId.toString(), 'chat:read', {
        chatId: chat._id,
        readBy: userId,
        readAt: new Date(),
      });
    }

    return updated;
  }

  /**
   * Get user's chat list
   */
  async getUserChats(
    userId: string,
    userRole: string,
    page: number = 1,
    limit: number = 20
  ): Promise<{
    chats: IChat[];
    total: number;
    page: number;
    pages: number;
  }> {
    let query: Record<string, unknown> = {};

    if (userRole === 'customer') {
      const customer = await Customer.findOne({ userId });
      if (!customer) {
        throw new NotFoundError('العميل غير موجود');
      }
      query = { 'participants.customerId': customer._id };
    } else if (userRole === 'restaurant') {
      const restaurant = await Restaurant.findOne({ userId });
      if (!restaurant) {
        throw new NotFoundError('المطعم غير موجود');
      }
      query = { 'participants.restaurantId': restaurant._id };
    } else if (userRole === 'driver' || userRole === 'delivery') {
      const driver = await Driver.findOne({ userId });
      if (!driver) {
        throw new NotFoundError('السائق غير موجود');
      }
      query = { 'participants.driverId': driver._id };
    }

    const total = await Chat.countDocuments(query);
    const pages = Math.ceil(total / limit);
    const skip = (page - 1) * limit;

    const chats = await Chat.find(query)
      .populate('orderId', 'orderNumber status')
      .sort({ lastMessageAt: -1 })
      .skip(skip)
      .limit(limit);

    return {
      chats,
      total,
      page,
      pages,
    };
  }

  /**
   * Get chat by order ID
   */
  async getChatByOrder(
    orderId: string,
    chatType: 'customer_restaurant' | 'customer_driver' | 'restaurant_driver',
    userId: string,
    userRole: string
  ): Promise<IChat | null> {
    const chat = await Chat.findOne({ orderId, chatType });

    if (chat) {
      // Mark as read when accessed
      await chat.markAsRead(new mongoose.Types.ObjectId(userId));
    }

    return chat;
  }

  /**
   * Deactivate chat (usually when order is completed/cancelled)
   */
  async deactivateChat(orderId: string): Promise<void> {
    await Chat.updateMany(
      { orderId },
      { isActive: false }
    );
  }

  /**
   * Get unread count for user
   */
  async getUnreadCount(userId: string, userRole: string): Promise<number> {
    let query: Record<string, unknown> = {};

    if (userRole === 'customer') {
      const customer = await Customer.findOne({ userId });
      if (!customer) return 0;
      query = { 'participants.customerId': customer._id };
    } else if (userRole === 'restaurant') {
      const restaurant = await Restaurant.findOne({ userId });
      if (!restaurant) return 0;
      query = { 'participants.restaurantId': restaurant._id };
    } else if (userRole === 'driver' || userRole === 'delivery') {
      const driver = await Driver.findOne({ userId });
      if (!driver) return 0;
      query = { 'participants.driverId': driver._id };
    }

    const chats = await Chat.find({ ...query, isActive: true });

    let totalUnread = 0;
    for (const chat of chats) {
      totalUnread += chat.getUnreadCount(new mongoose.Types.ObjectId(userId));
    }

    return totalUnread;
  }

  /**
   * Verify user is a participant in the chat type
   */
  private async verifyParticipant(
    order: typeof Order.prototype,
    userId: string,
    userRole: string,
    chatType: 'customer_restaurant' | 'customer_driver' | 'restaurant_driver'
  ): Promise<boolean> {
    if (userRole === 'customer') {
      const customer = await Customer.findOne({ userId });
      if (!customer) return false;
      return customer._id.toString() === order.customerId.toString() &&
        (chatType === 'customer_restaurant' || chatType === 'customer_driver');
    }

    if (userRole === 'restaurant') {
      const restaurant = await Restaurant.findOne({ userId });
      if (!restaurant) return false;
      return restaurant._id.toString() === order.restaurantId.toString() &&
        (chatType === 'customer_restaurant' || chatType === 'restaurant_driver');
    }

    if (userRole === 'driver' || userRole === 'delivery') {
      const driver = await Driver.findOne({ userId });
      if (!driver || !order.driverId) return false;
      return driver.userId.toString() === order.driverId.toString() &&
        (chatType === 'customer_driver' || chatType === 'restaurant_driver');
    }

    return false;
  }
}

export const chatService = new ChatService();
export default chatService;

File: customer-app/lib/services/chat_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'api_service.dart';

final chatServiceProvider = Provider<ChatService>((ref) {
  final apiService = ref.watch(apiServiceProvider);
  return ChatService(apiService);
});

class ChatService {
  final ApiService _apiService;

  ChatService(this._apiService);

  /// Get user's chat list
  Future<ChatsListResponse> getChats({
    int page = 1,
    int limit = 20,
  }) async {
    try {
      final response = await _apiService.get(
        '/chats',
        queryParameters: {
          'page': page,
          'limit': limit,
        },
      );

      if (response.data['success']) {
        return ChatsListResponse.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'فشل جلب المحادثات');
      }
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'حدث خطأ أثناء جلب المحادثات',
        );
      } else {
        throw Exception('لا يوجد اتصال بالإنترنت');
      }
    }
  }

  /// Get or create chat for an order
  Future<Chat> getOrCreateChat({
    required String orderId,
    required String chatType, // 'customer_restaurant' or 'customer_driver'
  }) async {
    try {
      final response = await _apiService.get(
        '/chats/order/$orderId',
        queryParameters: {
          'chatType': chatType,
        },
      );

      if (response.data['success']) {
        return Chat.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'فشل جلب المحادثة');
      }
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'حدث خطأ أثناء جلب المحادثة',
        );
      } else {
        throw Exception('لا يوجد اتصال بالإنترنت');
      }
    }
  }

  /// Get chat messages
  Future<MessagesResponse> getMessages({
    required String chatId,
    int page = 1,
    int limit = 50,
  }) async {
    try {
      final response = await _apiService.get(
        '/chats/$chatId/messages',
        queryParameters: {
          'page': page,
          'limit': limit,
        },
      );

      if (response.data['success']) {
        return MessagesResponse.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'فشل جلب الرسائل');
      }
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'حدث خطأ أثناء جلب الرسائل',
        );
      } else {
        throw Exception('لا يوجد اتصال بالإنترنت');
      }
    }
  }

  /// Send a message
  Future<ChatMessage> sendMessage({
    required String chatId,
    required String content,
    String type = 'text',
    String? imageUrl,
    double? lat,
    double? lng,
  }) async {
    try {
      final Map<String, dynamic> data = {
        'content': content,
        'type': type,
      };

      if (imageUrl != null) {
        data['imageUrl'] = imageUrl;
      }

      if (lat != null && lng != null) {
        data['location'] = {
          'lat': lat,
          'lng': lng,
        };
      }

      final response = await _apiService.post(
        '/chats/$chatId/messages',
        data: data,
      );

      if (response.data['success']) {
        return ChatMessage.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'فشل إرسال الرسالة');
      }
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'حدث خطأ أثناء إرسال الرسالة',
        );
      } else {
        throw Exception('لا يوجد اتصال بالإنترنت');
      }
    }
  }

  /// Mark chat as read
  Future<void> markAsRead(String chatId) async {
    try {
      await _apiService.put('/chats/$chatId/read');
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'حدث خطأ',
        );
      } else {
        throw Exception('لا يوجد اتصال بالإنترنت');
      }
    }
  }

  /// Get unread count
  Future<int> getUnreadCount() async {
    try {
      final response = await _apiService.get('/chats/unread-count');

      if (response.data['success']) {
        return response.data['data']['unreadCount'] as int;
      } else {
        return 0;
      }
    } catch (_) {
      return 0;
    }
  }

  /// Upload image for chat
  Future<String> uploadImage(String filePath) async {
    try {
      final formData = FormData.fromMap({
        'file': await MultipartFile.fromFile(filePath),
        'folder': 'chats',
      });

      final response = await _apiService.post(
        '/upload/image',
        data: formData,
      );

      if (response.data['success']) {
        return response.data['data']['url'] as String;
      } else {
        throw Exception(response.data['message'] ?? 'فشل رفع الصورة');
      }
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'حدث خطأ أثناء رفع الصورة',
        );
      } else {
        throw Exception('لا يوجد اتصال بالإنترنت');
      }
    }
  }
}

/// Chat Model
class Chat {
  final String id;
  final String orderId;
  final String chatType;
  final ChatParticipants participants;
  final String? lastMessage;
  final DateTime? lastMessageAt;
  final String? lastMessageBy;
  final bool isActive;
  final int unreadCount;
  final OrderInfo? order;

  Chat({
    required this.id,
    required this.orderId,
    required this.chatType,
    required this.participants,
    this.lastMessage,
    this.lastMessageAt,
    this.lastMessageBy,
    required this.isActive,
    this.unreadCount = 0,
    this.order,
  });

  factory Chat.fromJson(Map<String, dynamic> json) {
    return Chat(
      id: json['_id'] as String,
      orderId: json['orderId'] is Map
          ? json['orderId']['_id'] as String
          : json['orderId'] as String,
      chatType: json['chatType'] as String,
      participants: ChatParticipants.fromJson(
        json['participants'] as Map<String, dynamic>,
      ),
      lastMessage: json['lastMessage'] as String?,
      lastMessageAt: json['lastMessageAt'] != null
          ? DateTime.parse(json['lastMessageAt'] as String)
          : null,
      lastMessageBy: json['lastMessageBy'] as String?,
      isActive: json['isActive'] as bool? ?? true,
      unreadCount: json['unreadCount'] as int? ?? 0,
      order: json['orderId'] is Map
          ? OrderInfo.fromJson(json['orderId'] as Map<String, dynamic>)
          : null,
    );
  }

  String get chatTypeLabel {
    switch (chatType) {
      case 'customer_restaurant':
        return 'محادثة مع المطعم';
      case 'customer_driver':
        return 'محادثة مع السائق';
      default:
        return 'محادثة';
    }
  }
}

/// Chat Participants
class ChatParticipants {
  final String? customerId;
  final String? restaurantId;
  final String? driverId;

  ChatParticipants({
    this.customerId,
    this.restaurantId,
    this.driverId,
  });

  factory ChatParticipants.fromJson(Map<String, dynamic> json) {
    return ChatParticipants(
      customerId: json['customerId'] as String?,
      restaurantId: json['restaurantId'] as String?,
      driverId: json['driverId'] as String?,
    );
  }
}

/// Order Info (for chat list)
class OrderInfo {
  final String id;
  final String orderNumber;
  final String status;

  OrderInfo({
    required this.id,
    required this.orderNumber,
    required this.status,
  });

  factory OrderInfo.fromJson(Map<String, dynamic> json) {
    return OrderInfo(
      id: json['_id'] as String,
      orderNumber: json['orderNumber'] as String? ?? '',
      status: json['status'] as String? ?? '',
    );
  }
}

/// Chat Message Model
class ChatMessage {
  final String? id;
  final String sender;
  final String senderRole;
  final String content;
  final String type;
  final String? imageUrl;
  final MessageLocation? location;
  final bool isRead;
  final DateTime createdAt;
  final bool isSending;

  ChatMessage({
    this.id,
    required this.sender,
    required this.senderRole,
    required this.content,
    this.type = 'text',
    this.imageUrl,
    this.location,
    this.isRead = false,
    required this.createdAt,
    this.isSending = false,
  });

  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    return ChatMessage(
      id: json['_id'] as String?,
      sender: json['sender'] as String,
      senderRole: json['senderRole'] as String,
      content: json['content'] as String,
      type: json['type'] as String? ?? 'text',
      imageUrl: json['imageUrl'] as String?,
      location: json['location'] != null
          ? MessageLocation.fromJson(json['location'] as Map<String, dynamic>)
          : null,
      isRead: json['isRead'] as bool? ?? false,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  ChatMessage copyWith({
    String? id,
    String? sender,
    String? senderRole,
    String? content,
    String? type,
    String? imageUrl,
    MessageLocation? location,
    bool? isRead,
    DateTime? createdAt,
    bool? isSending,
  }) {
    return ChatMessage(
      id: id ?? this.id,
      sender: sender ?? this.sender,
      senderRole: senderRole ?? this.senderRole,
      content: content ?? this.content,
      type: type ?? this.type,
      imageUrl: imageUrl ?? this.imageUrl,
      location: location ?? this.location,
      isRead: isRead ?? this.isRead,
      createdAt: createdAt ?? this.createdAt,
      isSending: isSending ?? this.isSending,
    );
  }
}

/// Message Location
class MessageLocation {
  final double lat;
  final double lng;

  MessageLocation({
    required this.lat,
    required this.lng,
  });

  factory MessageLocation.fromJson(Map<String, dynamic> json) {
    return MessageLocation(
      lat: (json['lat'] as num).toDouble(),
      lng: (json['lng'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'lat': lat,
      'lng': lng,
    };
  }
}

/// Chats List Response
class ChatsListResponse {
  final List<Chat> chats;
  final int total;
  final int page;
  final int pages;

  ChatsListResponse({
    required this.chats,
    required this.total,
    required this.page,
    required this.pages,
  });

  factory ChatsListResponse.fromJson(Map<String, dynamic> json) {
    final List<dynamic> chatsData = json['chats'] as List<dynamic>;

    return ChatsListResponse(
      chats: chatsData
          .map((c) => Chat.fromJson(c as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int? ?? 0,
      page: json['page'] as int? ?? 1,
      pages: json['pages'] as int? ?? 1,
    );
  }
}

/// Messages Response
class MessagesResponse {
  final List<ChatMessage> messages;
  final int total;
  final int page;
  final int pages;

  MessagesResponse({
    required this.messages,
    required this.total,
    required this.page,
    required this.pages,
  });

  factory MessagesResponse.fromJson(Map<String, dynamic> json) {
    final List<dynamic> messagesData = json['messages'] as List<dynamic>;

    return MessagesResponse(
      messages: messagesData
          .map((m) => ChatMessage.fromJson(m as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int? ?? 0,
      page: json['page'] as int? ?? 1,
      pages: json['pages'] as int? ?? 1,
    );
  }
}
